tests/pass/float.rs

```diff
@@ -64,13 +105,59 @@ float_to_int!(f64 => i8, u8, i16, u16, i32, u32, i64, u64, i128, u128);
 /// Test this cast both via `as` and via `approx_unchecked` (i.e., it must not saturate).
 #[track_caller]
 #[inline(never)]
-fn test_both_cast<F, I>(x: F, y: I)
+fn test_both_cast<F, I>(x: F, y: I, msg: &str)
 where
     F: FloatToInt<I>,
     I: PartialEq + Debug,
 {
-    assert_eq!(x.cast(), y);
-    assert_eq!(unsafe { x.cast_unchecked() }, y);
+    assert_eq!(x.cast(), y, "{} -> {}: {msg}", type_name::<F>(), type_name::<I>());
+    assert_eq!(
+        unsafe { x.cast_unchecked() },
+        y,
+        "{} -> {}: {msg}",
+        stringify!($fty),
+        stringify!($ity)
+    );
+}
+
+/// Helper function to avoid promotion so that this tests "run-time" casts, not CTFE.
+/// Doesn't make a big difference when running this in Miri, but it means we can compare this
+/// with the LLVM backend by running `rustc -Zmir-opt-level=0 -Zsaturating-float-casts`.
+#[track_caller]
+#[inline(never)]
+fn assert_eq<T: PartialEq + Debug>(x: T, y: T) {
+    assert_eq!(x, y);
+}
+
+/// The same as `assert_eq` except prints a specific message on failure
+#[track_caller]
+#[inline(never)]
+fn assert_eq_msg<T: PartialEq + Debug>(x: T, y: T, msg: &str) {
+    assert_eq!(x, y, "{msg}");
+}
+
+/// Check that floats have bitwise equality
+fn assert_biteq<F: Float>(a: F, b: F, msg: &str) {
+    let ab = a.to_bits();
+    let bb = b.to_bits();
+    let msg = format!(
+        "({ab:#0width$x} != {bb:#0width$x}) {}: {msg}",
+        type_name::<F>(),
+        width = (2 + F::Int::BITS / 4) as usize
+    );
+    assert_eq_msg::<F::Int>(ab, bb, &msg);
+}
+
+/// Check that two floats have equality
+fn assert_feq<F: Float>(a: F, b: F, msg: &str) {
+    let ab = a.to_bits();
+    let bb = b.to_bits();
+    let msg = format!(
```

Comment:
There's a lot of `format!` happening for strings that will never be shown... I guess even Miri is fast enough for this but I still find it mildly displeasing.^^

Does it work to make `assert_eq_msg` take `impl Display` or so?

Reply:
That does indeed work, updated